jolt-core/src/subprotocols/ra_virtual.rs

```diff
@@ -0,0 +1,204 @@
+                poly.bind_parallel(r_value, BindingOrder::LowToHigh);
+            }
+        }
+
+        Self {
+            ra_claim,
+            ra_i_polys,
+            eq_poly,
+            r_cycle,
+            r_address,
+            r_address_chunks,
+            ra_i_claims: None,
+            T,
+        }
+    }
+
+    pub fn prove<ProofTranscript: Transcript>(
+        mut self,
+        transcript: &mut ProofTranscript,
+    ) -> (RAProof<F, ProofTranscript>, Vec<F>) {
+        let (sumcheck_proof, r_cycle_bound) =
+            crate::subprotocols::sumcheck::BatchedSumcheck::prove(vec![&mut self], transcript);
+
+        let ra_i_claims = self
+            .ra_i_claims
+            .expect("ra_i_claims were not set after prove")
+            .to_vec();
+
+        let proof = RAProof {
+            sumcheck_proof,
+            ra_i_claims,
+        };
+
+        (proof, r_cycle_bound)
+    }
+}
+
+impl<F: JoltField, ProofTranscript: Transcript, const D: usize>
+    BatchableSumcheckInstance<F, ProofTranscript> for RASumcheck<F, D>
+{
+    fn degree(&self) -> usize {
+        D + 1
+    }
+
+    fn num_rounds(&self) -> usize {
+        self.T.log_2()
+    }
+
+    fn cache_openings(&mut self) {
+        let mut openings = [F::zero(); D];
+
+        for i in 0..D {
+            openings[i] = self.ra_i_polys[i].final_sumcheck_claim();
+        }
+
+        self.ra_i_claims = Some(openings);
+    }
+
+    fn bind(&mut self, r_j: F, _: usize) {
+        for ra_i in self.ra_i_polys.iter_mut() {
+            ra_i.bind_parallel(r_j, BindingOrder::LowToHigh);
+        }
+        self.eq_poly.bind_parallel(r_j, BindingOrder::LowToHigh);
+    }
+
+    fn input_claim(&self) -> F {
+        self.ra_claim.clone()
+    }
+
+    fn expected_output_claim(&self, r: &[F]) -> F {
+        
+        let eq_eval = self.eq_poly.evaluate(r);
+        
+        // Compute the product of all ra_i evaluations
+        let mut product = F::one();
+        for (_i, ra_i_claim) in self.ra_i_claims.as_ref().unwrap().iter().enumerate() {
+            product *= *ra_i_claim;
+        }
+        
+        eq_eval * product
+    }
+
+    fn compute_prover_message(&self, round: usize) -> Vec<F> {
+        let degree = <Self as BatchableSumcheckInstance<F, ProofTranscript>>::degree(self);
+        let ra_i_polys = &self.ra_i_polys;
+        let eq_poly = &self.eq_poly;
+        
+        // We need to compute evaluations at 0, 2, 3, ..., degree
+        // = eq(r_cycle, j) * ∏_{i=0}^{D-1} ra_i(j)
+        
+        let eval_points: Vec<usize> = (0..=degree)
+            .filter(|&i| i != 1)
+            .collect();
+        
+        let mut evals = vec![F::zero(); eval_points.len()];
+
+        let remaining_vars = <Self as BatchableSumcheckInstance<F, ProofTranscript>>::num_rounds(&self) - round - 1;
+        
+        for (eval_idx, &point) in eval_points.iter().enumerate() {
+            for j in 0..(1 << remaining_vars) {
```

Comment:
I would invert these loops; see [here](https://github.com/a16z/jolt/blob/feat/twist-shout-integration/jolt-core/src/subprotocols/shout.rs#L171-L186) for the idiomatic way to implement this function